code/hash/Hacl.Hash.Definitions.fst

```diff
@@ -49,45 +97,63 @@ let hash_len (a: hash_alg): n:size_t { v n = hash_length a } =
   | SHA2_256 -> 32ul
   | SHA2_384 -> 48ul
   | SHA2_512 -> 64ul
+  | Blake2S -> 32ul
+  | Blake2B -> 64ul
 
 noextract inline_for_extraction
 let blocks_t (a: hash_alg) =
   b:B.buffer uint8 { B.length b % block_length a = 0 }
 
 let hash_t (a: hash_alg) = b:B.buffer uint8 { B.length b = hash_length a }
 
+// The proper way to generate an extra state from a constant nat.
+noextract inline_for_extraction
+let const_nat_to_extra_state (a:hash_alg{is_blake a}) (n:nat{n <= maxint U64}) :
+  extra_state a =
+  match a with
+  | Blake2S -> mk_int #U64 #SEC n
+  | Blake2B -> cast U128 SEC (mk_int #U64 #SEC n)
+
+noextract inline_for_extraction
+let initial_extra_state (a:hash_alg) : extra_state a =
+  if is_blake a then const_nat_to_extra_state a 0
+  else ()
 
 (** The types of all stateful operations for a hash algorithm. *)
 
 noextract inline_for_extraction
-let alloca_st (a: hash_alg) = unit -> ST.StackInline (state a)
+let alloca_st (i:impl) = unit -> ST.StackInline (state i & extra_state (get_alg i))
   (requires (fun h ->
     HS.is_stack_region (HS.get_tip h)))
-  (ensures (fun h0 s h1 ->
+  (ensures (fun h0 (s, v) h1 ->
     M.(modifies M.loc_none h0 h1) /\
     B.live h1 s /\
     B.frameOf s == HS.get_tip h0 /\
-    Seq.equal (B.as_seq h1 s) (Spec.Agile.Hash.init a) /\
-    LowStar.Monotonic.Buffer.alloc_post_mem_common s h0 h1 (Spec.Agile.Hash.init a)))
+    (as_seq h1 s, v) == Spec.Agile.Hash.init (get_alg i) /\
```

Comment:
this seems overly precise; where did you get the inspiration for this post-condition? the equality of domains should be given to you already by StackInline, same thing for the equality of the tips, and the `unused_in` should probably be replaced by `fresh_loc`